src/libs/Middleware/Reauthentication.js

```diff
@@ -0,0 +1,88 @@
+import lodashGet from 'lodash/get';
+import CONST from '../../CONST';
+import * as NetworkStore from '../Network/NetworkStore';
+import * as Network from '../Network';
+import * as Authentication from '../Authentication';
+import * as PersistedRequests from '../actions/PersistedRequests';
+import * as Request from '../Request';
+
+/**
+ * Reauthentication middleware
+ *
+ * @param {Promise} response
+ * @param {Object} request
+ * @param {Boolean} isFromSequentialQueue
+ * @returns {Promise}
+ */
+function Reauthentication(response, request, isFromSequentialQueue) {
+    return response
+        .then((data) => {
+            if (data.jsonCode === CONST.JSON_CODE.NOT_AUTHENTICATED) {
+                const credentials = NetworkStore.getCredentials();
+
+                // Credentials haven't been initialized. We will not be able to re-authenticate with the API.
+                const unableToReauthenticate = (!credentials || !credentials.autoGeneratedLogin || !credentials.autoGeneratedPassword);
+                if (unableToReauthenticate) {
+                    if (isFromSequentialQueue) {
+                        throw new Error('Missing credentials required for authentication');
+                    }
+
+                    Network.replayRequest(request);
+                    return;
+                }
+
+                // There are some API requests that should not be retried when there is an auth failure like
+                // creating and deleting logins. In those cases, they should handle the original response instead
+                // of the new response created by handleExpiredAuthToken.
+                const shouldRetry = lodashGet(request, 'data.shouldRetry');
+                if (!shouldRetry) {
+                    if (isFromSequentialQueue) {
+                        return data;
+                    }
+
+                    request.resolve(data);
+                    return;
+                }
+
+                // We are already authenticating
+                if (NetworkStore.getIsAuthenticating()) {
```

Comment:
What do you think about moving this to `Authenticate.isAuthenticating()`?

Reply:
One idea I am having is to create an `AuthenticationUtils`. 

I've tried a couple different things and didn't land on how to fix the dependency cycle (I need more practice fixing those puzzles 😄 ), but it maybe is reasonable to have `authenticate` and `reauthenticate` in `Authentication` and put the credentials and isAuthenticating stuff in an `AuthenticationUtils`

Reply:
Ah, circular dependencies can be a huge pain like that. I think I'd rather leave them on `NetworkStore` rather than having a new `AuthenticationUtils`.

It's at least better that `Network` doesn't have any specialized knowledge now, but it is a little hard trying to reason about the difference between `Network` and `NetworkStore`.